From ef9563c799b8913f6bc6233142e42bc6244d5556 Mon Sep 17 00:00:00 2001
From: Jian He <jianhe@apache.org>
Date: Thu, 13 Aug 2015 16:02:57 -0700
Subject: [PATCH 1385/2848] YARN-4047. ClientRMService getApplications has
 high scheduler lock contention. Contributed by
 Jason Lowe (cherry picked from commit
 7a445fcfabcf9c6aae219051f65d3f6cb8feb87c)

Conflicts:
	hadoop-yarn-project/CHANGES.txt

(cherry picked from commit de1c4121a1e70bab968d65b8ee9f8b62e54ba639)

Change-Id: Idf6ccd9f9769411a98ea13117417e7098817f2e2
---
 .../server/resourcemanager/ClientRMService.java    |   11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ClientRMService.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ClientRMService.java
index 83f37ec..3e9ab0c 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ClientRMService.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/ClientRMService.java
@@ -745,13 +745,9 @@ public void remove() {
       RMApp application = appsIter.next();
 
       // Check if current application falls under the specified scope
-      boolean allowAccess = checkAccess(callerUGI, application.getUser(),
-          ApplicationAccessType.VIEW_APP, application);
       if (scope == ApplicationsRequestScope.OWN &&
           !callerUGI.getUserName().equals(application.getUser())) {
         continue;
-      } else if (scope == ApplicationsRequestScope.VIEWABLE && !allowAccess) {
-        continue;
       }
 
       if (applicationTypes != null && !applicationTypes.isEmpty()) {
@@ -800,6 +796,13 @@ public void remove() {
         }
       }
 
+      // checkAccess can grab the scheduler lock so call it last
+      boolean allowAccess = checkAccess(callerUGI, application.getUser(),
+          ApplicationAccessType.VIEW_APP, application);
+      if (scope == ApplicationsRequestScope.VIEWABLE && !allowAccess) {
+        continue;
+      }
+
       reports.add(application.createAndGetApplicationReport(
           callerUGI.getUserName(), allowAccess));
     }
-- 
1.7.9.5

